Packages/ClientRuntime/Sources/Waiters/WaiterUtils.swift

```diff
@@ -0,0 +1,120 @@
+        // the compiler selects an overload.
+        // Without the unused argument, the compiler uses this implementation instead of the one above when
+        // the element of an array is also an array, which is not desired behavior.
+        // See: https://forums.swift.org/t/how-to-specify-which-is-the-default-function-when-functions-use-the-same-name-like-array-reversed/39168/5
+        // In the future, if Swift allows us to manually rank overloads or otherwise fixes this issue,
+        // we should eliminate this param & use the new mechanism for overload resolution instead.
+        return try map { try transform($0) }
+    }
+}
+
+/// A value type that is used to facilitate comparison of values returned by JMESPath types, in accordance with the JMESPath spec.
+public enum JMESValue: Equatable, Comparable {
+    /// A value that is a "number" per the JMESPath spec.
+    case number(Double)
+    /// A value that is a "boolean" per the JMESPath spec.
+    case boolean(Bool)
+    /// A value that is a "string" per the JMESPath spec.
+    case string(String)
+    /// A value that is "null" per the JMESPath spec.
+    case null
+
+    /// Creates a `JMESValue` from a Swift `Int`.
+    /// - Parameter int: The `Int` to represent as a JMESPath number, or as a JMESPath `null` if `int` is nil.
+    public init(_ int: Int?) {
+        guard let int = int else { self = .null; return }
+        self = .number(Double(int))
+    }
+
+    /// Creates a `JMESValue` from a Swift `Double`.
+    /// - Parameter double: The `Double` to represent as a JMESPath number, or as a JMESPath `null` if `double` is nil.
+    public init(_ double: Double?) {
+        guard let double = double else { self = .null; return }
+        self = .number(double)
+    }
+
+    /// Creates a `JMESValue` from a Swift `Bool`.
+    /// - Parameter bool: The `Bool` to represent as a JMESPath Boolean, or as a JMESPath `null` if `bool` is nil.
+    public init(_ bool: Bool?) {
+        guard let bool = bool else { self = .null; return }
+        self = .boolean(bool)
+    }
+
+    /// Creates a `JMESValue` from a Swift `String`.
+    /// - Parameter string: The `String` to represent as a JMESPath String, or as a JMESPath `null` if `string` is nil.
+    public init(_ string: String?) {
+        guard let string = string else { self = .null; return }
+        self = .string(string)
+    }
+
+    /// Creates a `JMESValue` from a type that is raw-representable by a string.
+    /// Typically this will be an enumeration of a limited set of expected string values.
+    /// The resulting value will be a string for purposes of comparing it to other values.
+    /// - Parameter rr: The `RawRepresentable` to represent as a JMESPath String, or as a JMESPath `null` if `rr` is nil.
+    public init<T: RawRepresentable>(_ rr: T?) where T.RawValue == String {
+        guard let string = rr?.rawValue else { self = .null; return }
+        self = .string(string)
+    }
```

Comment:
Can we call the string initializer to simplify things? Additionally, prefer to use a clearly named parameter over abbreviations. 

```suggestion
    public init<T: RawRepresentable>(_ value: T?) where T.RawValue == String {
        self.init(value?.rawValue)
    }
```